testsuite/classlibrary/TestProxySpace.sc

```diff
@@ -0,0 +1,62 @@
+TestProxySpace : UnitTest {
+
+
+	test_storeOn_recoversState {
+		var psA, psB, stream, bValue, bNodeMap, failedKey;
+		var server = Server(\test, NetAddr("127.0.0.1", 57111));
+
+		this.bootServer(server);
+		server.initTree;
+
+		psA = ProxySpace(server).make {
+			~out = { \in.ar(0!2) };
+			~osc = { |freq = 440| SinOsc.ar(freq).dup };
+			~out.set(\in, ~osc);
+			~freqmod = { LFDNoise3.kr(14).exprange(200, 800) };
+			~osc.set(\freq, ~freqmod);
+
+			// must do this inside the ProxySpace environment
+			// otherwise it can't identify the ~keys
+			stream = CollStream.new;
+			currentEnvironment.storeOn(stream);
+		};
+
+		psB = ProxySpace(server).make {
+			stream.collection.interpret;
+		};
+
+		failedKey = block { |break|
+			psA.keysValuesDo { |key, value|
+				bValue = psB[key];
+				// basic checks
+				if(bValue.rate != value.rate or: {
+					bValue.numChannels != value.numChannels or: {
+						bValue.source.asCompileString != value.source.asCompileString
+					}
+				}) {
+					break.value(key);
+				};
+				// check nodemap
+				bNodeMap = bValue.nodeMap;
+				value.nodeMap.keysValuesDo { |parm, map|
+					// output bus indices will not match, skip them
+					if(#[out, i_out].includes(parm).not) {
+						if(map.asCompileString != bNodeMap[parm].asCompileString) {
+							break.value(key);
+						};
+					};
+				};
+			};
+			nil
+		};
+
+		psA.clear;
+		psB.clear;
+		server.quit;
+
+		this.assert(failedKey.isNil,
+			"ProxySpace document string should recover all proxies' sources, rates, numChannels and nodemaps %"
+			.format(if(failedKey.notNil) { "(% failed)".format(failedKey) } { "" })
+		);
```

Comment:
why all this extra effort with `failedKey` and `block {}`? just put the asserts inside the `keysValuesDo` loop. here are a few reasons why that would be better:
- test message can include more details on the specific failure (with the `details:` parameter, which you should be using here)
- no need to have this `if(failedKey.notNil)` logic (line 59)
- will be able to see all failures if multiple cases are failing
- simpler and clearer

you should also break apart this check:
```supercollider
if(bValue.rate != value.rate or: {
	bValue.numChannels != value.numChannels or: {
		bValue.source.asCompileString != value.source.asCompileString
	}
}) {
```

into

```supercollider
assertEquals(bValue.rate, value.rate, "message", details: "details");
assertEquals(bValue.numChannels, value.numChannels, "message", details: "details");
assertEquals(bValue.source.asCompileString, value.source.asCompileString, "message", details: "details");
```

so no information is hidden

Reply:
OK, I will change this.

I was thinking of the point in the unit testing guide, "Tests should have exactly one assertion." https://github.com/supercollider/supercollider/wiki/Unit-Testing-Guide#tests-should-have-exactly-one-assertion Is this point now out of date?

Reply:
Oh, and I (perhaps mistakenly) based this PR on 3.10, where `assertEquals` doesn't have a `details` argument. Should I close this and redo it against a different branch? Which branch?

Reply:
I think I did this ages ago. Resolve this review now?

Also I still think that the unit testing guide needs to be updated. I've seen several unit tests since this where multiple asserts were recommended, but the unit testing guide still lists this as a no-no.